proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,558 @@
+Safe Fixed Size Buffers
+=====
+
+## Summary
+
+Provide a general-purpose and safe mechanism for declaring fixed sized buffers within C# classes, structs, and interfaces.
+
+## Motivation
+
+This proposal plans to address the many limitations of https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/unsafe-code.md#228-fixed-size-buffers.
+Specifically it aims to allow the declaration of safe `fixed` buffers for managed and unmanaged types in a `struct`, `class`, or `interface`, and provide language safety verification for them.
+
+Note, that for the purpose of this proposal a term "fixed-size buffer" refers to a "safe fixed-size buffer" rather than to a buffer described at https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/unsafe-code.md#228-fixed-size-buffers.
+
+## Detailed Design (Option 1)
+
+The grammar for *variable_declarator* in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/classes.md#145-fields
+will be extended to allow specifying the size of the buffer:
+
+``` diff antlr
+field_declaration
+    : attributes? field_modifier* type variable_declarators ';'
+    ;
+
+field_modifier
+    : 'new'
+    | 'public'
+    | 'protected'
+    | 'internal'
+    | 'private'
+    | 'static'
+    | 'readonly'
+    | 'volatile'
+    | unsafe_modifier   // unsafe code support
+    ;
+
+variable_declarators
+    : variable_declarator (',' variable_declarator)*
+    ;
+    
+variable_declarator
+    : identifier ('=' variable_initializer)?
++   | fixed_size_buffer_declarator
+    ;
+    
+fixed_size_buffer_declarator
+    : identifier '[' constant_expression ']'
+    ;    
+```
+
+A *fixed_size_buffer_declarator* introduces a new field and consists of an identifier that names it,
+followed by a constant expression enclosed in `[` and `]` tokens. The type of the constant expression must be
+implicitly convertible to type `int`, and the value must be a non-zero positive integer.
+
+A *field_declaration* with a *fixed_size_buffer_declarator* in an interface must have `static` modifier.
+
+The field has a special struct type which represents a fixed-size buffer of a given element type.
+The element type is the *type* specified in *field_declaration*.  The constant expression
+denotes the number of elements in the buffer. The elements of a fixed-size buffer are laid out sequentially
+in memory as though they are elements of an array.
+
+The struct type will be defined using [```System.Runtime.CompilerServices.InlineArrayAttribute```](https://github.com/dotnet/runtime/issues/61135) feature.
+The framework will likely include a set of predefined "buffer" types that cover a limited set of buffer sizes. When a predefined type doesn't exist,
+compiler will synthesize it in the module being built. Names of the generated types will be "speakable" in order to support consumption from other languages. 
```

Comment:
I think we should reconsider this and not have them be speakable.  Related to Jan's feedback about System.Runtime.CompilerServices, I think we should instead:
- Have the compiler recognize whatever known public types we choose to expose from System.Runtime.CompilerServices
- Generate anything else it needs with internal unspeakable types.

This avoids conflicts with any types the user might declare, and avoids encouraging users to create types in System.Runtime.CompilerServices, which in general we don't want them doing.  I know and appreciate that one of the motivations for this was the libraries team creating those few built-in public helpers, but at most there will only be a handful of those and we can write them manually (even copy/pasting/tweaking what the compiler generated).

I also want the compiler to be free to add additional surface area to these types in the future.  For one of the same reason we don't publicly expose records, if the compiler is free to add additional APIs in the future, we wouldn't use it to create these public types, anyway.